Sources/Basic/FileInfo.swift

```diff
@@ -8,71 +8,37 @@
  See http://swift.org/CONTRIBUTORS.txt for Swift project authors
  */
 
-import SPMLibc
 import Foundation
 
 /// File system information for a particular file.
 public struct FileInfo: Equatable, Codable {
 
-    /// File timestamp wrapper.
-    public struct FileTimestamp: Equatable, Codable {
-        public let seconds: UInt64
-        public let nanoseconds: UInt64
-    }
-
-    /// File system entity kind.
-    public enum Kind: String, Codable {
-        case file, directory, symlink, blockdev, chardev, socket, unknown
-
-        fileprivate init(mode: mode_t) {
-            switch mode {
-            case S_IFREG:  self = .file
-            case S_IFDIR:  self = .directory
-            case S_IFLNK:  self = .symlink
-            case S_IFBLK:  self = .blockdev
-            case S_IFCHR:  self = .chardev
-            case S_IFSOCK: self = .socket
-            default:
-                self = .unknown
-            }
-        }
-    }
-
-    /// The device number.
-    public let device: UInt64
+        /// The device number.
+    public let device: UInt64?
 
     /// The inode number.
-    public let inode: UInt64
-
-    /// The mode flags of the file.
-    public let mode: UInt64
+    public let inode: UInt64?
 
     /// The size of the file.
-    public let size: UInt64
+    public let size: UInt64?
 
     /// The modification time of the file.
-    public let modTime: FileTimestamp
+    public let modTime: Date?
 
     /// Kind of file system entity.
-    public var kind: Kind {
-        return Kind(mode: mode_t(mode) & S_IFMT)
-    }
-
-    public init(_ buf: SPMLibc.stat) {
-        self.device = UInt64(buf.st_dev)
-        self.inode = UInt64(buf.st_ino)
-        self.mode = UInt64(buf.st_mode)
-        self.size = UInt64(buf.st_size)
-
-        #if canImport(Darwin)
-        let seconds = buf.st_mtimespec.tv_sec
-        let nanoseconds = buf.st_mtimespec.tv_nsec
-        #else
-        let seconds = buf.st_mtim.tv_sec
-        let nanoseconds = buf.st_mtim.tv_nsec
-        #endif
+    public let posixPermissions: Int16?
 
-        self.modTime = FileTimestamp(
-            seconds: UInt64(seconds), nanoseconds: UInt64(nanoseconds))
+    /// Kind of file system entity.
+    public let fileType: FileAttributeType?
+
+    public init(_ attrs: [FileAttributeKey : Any]) {
+        self.device = (attrs[.deviceIdentifier] as? NSNumber)?.uint64Value
+        self.inode = attrs[.systemFileNumber] as? UInt64
+        self.posixPermissions = (attrs[.posixPermissions] as? NSNumber)?.int16Value
+        self.fileType = attrs[.type] as? FileAttributeType
+        self.size = attrs[.size] as? UInt64
+        self.modTime = attrs[.modificationDate] as? Date
```

Comment:
I am a little concerned about this. Maybe put an assert so we at least get to know when these values are present?

Reply:
I thought the idea was that the values may not be present?
```
Lily Vulcano [9:42 AM]
Force-casting isn’t safe; the file system may not return those values for one reason or another.
```
It makes sense to return nil in that case, doesn't it? Else I suppose I could make them all !s and have it crash if they're not there and we can figure out why they aren't there.

Reply:
Assertions are not triggered in release builds. I don't think a lot will go wrong if we don't have this info but it'll be nice to catch issues in debug builds which is usually what SwiftPM developers themselves use.

Reply:
Alright, added in the asserts, and removed the optional typing of the fields.